includes/amp-helper-functions.php

```diff
@@ -1696,6 +1706,13 @@ function amp_get_content_sanitizers( $post = null ) {
 			);
 		}
 
+		// Mark the `Invalid URL protocol http:` for attribute src, action and href as being in dev mode.
+		$dev_mode_xpaths[] = sprintf(
+			/* translators: %1$s is the home URL */
+			'//*[contains(@src, "%1$s") or contains(@action, "%1$s") or contains(@href, "%1$s")][starts-with(@src, "http:") or starts-with(@action, "http:") or starts-with(@href, "http:")]',
```

Comment:
@westonruter Should we include `href` attribute as well?

Moreover, please note that the included elements are those that have attributes with the same origin URLs.

Reply:
The `href` attribute would have to be included, yes, as it needs to handle the case of the Web App Manifest `link`.

However, I think it should only include the `starts-with()` predicates and not `contains()`. 

Nevertheless, there won't address the case where there is a protocol-relative link. For example:

```html
<link rel="web-app-manifest" href="//localhost/sw.js">
```

Also, it wouldn't handle a relative path:

```html
<link rel="web-app-manifest" href="/sw.js">
```

Therefore, I'm wondering if marking with dev mode is going to work reliably.

Reply:
> However, I think it should only include the starts-with() predicates and not contains().

Then we will need some other workaround to filter out elements that have same-origin URLs in their attars. We still want to show an invalid URL protocol error if the URL is cross-origin... right? 

---

Makes more sense to update the sanitizer to handle the case. But there are two more cases:
1. Usage of localhost IP - `127.0.0.1`
2. Change in hostname using `/etc/hosts` mapping.       _// Should we even consider this?_

Reply:
> Then we will need some other workaround to filter out elements that have same-origin URLs in their attars. We still want to show an invalid URL protocol error if the URL is cross-origin... right?

What's an example of this attribute? Now that I think of it, `srcset` is one case in which there are multiple URLs in one attribute.

> Makes more sense to update the sanitizer to handle the case. But there are two more cases:

What we should follow here is what browsers allow for bypassing the HTTPS security requirement for certain web platform features. I know this includes `localhost` and `127.0.0.1`, but it is also true for `subdomain.localhost`. It would _not_ be true for custom hostnames defined in `/etc/hosts`. See more at https://web.dev/when-to-use-local-https/

Reply:
I mean a case something like this:
```html
// Should consider development mode
<amp-autocomplete filter="substring"
    src="http://localhost/static/samples/json/amp-autocomplete-cities.json">
    <input>
 </amp-autocomplete>
```
```html
// Maybe not consider it as development mode?
<amp-autocomplete filter="substring"
    src="http://example.com/static/samples/json/amp-autocomplete-cities.json">
    <input>
 </amp-autocomplete>

```

Reply:
Right, the exception should only be granted if: 

1. The page is being served from `localhost`, and
2. The attribute is linking to `localhost`.

Reply:
>  It would not be true for custom hostnames defined in /etc/hosts. See more at https://web.dev/when-to-use-local-https/

Perfect. Thanks for sharing. Also, I see `parse_url()` returns the host as `localhost` while using `127.0.0.1` or `[::1]`.

Reply:
@westonruter What about `amp_is_dev_mode` as new arg in `AMP_Tag_And_Attribute_Sanitizer`?

Reply:
How about something more specific, like `allow_localhost_http_protocol`